File: aimvision-backend/app/schemas/erasure.py
"""DTOs for the right-to-erasure endpoints."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ErasureRequestIn(BaseModel):
    athlete_user_id: str = Field(min_length=1, max_length=64)
    reason: str = Field(min_length=1, max_length=255)


class ErasureTicketOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    tenant_id: str
    athlete_user_id: str
    requested_by: str
    reason: str
    status: str
    references: dict[str, int] | None = Field(default=None, validation_alias="references_json")
    created_at: datetime
    completed_at: datetime | None = None

File: aimvision-mobile/src/services/erasure.ts
/**
 * Right-to-erasure (GDPR Art. 17 / ADR-0011) — mobile client.
 *
 * Two-step flow:
 *  1. `submitErasureRequest({athlete_user_id, reason})` — creates a
 *     pending ticket.
 *  2. `executeErasureTicket(ticketId)` — actually crypto-shreds the
 *     per-tenant DEK and marks the ticket complete. Separated so
 *     destructive action requires a confirmation step.
 */
import { api } from './api';

export interface ErasureTicket {
  id: string;
  tenant_id: string;
  athlete_user_id: string;
  requested_by: string;
  reason: string;
  status: 'pending' | 'completed' | 'failed' | string;
  references: Record<string, unknown> | null;
  created_at: string;
  completed_at: string | null;
}

export interface ErasureRequestIn {
  athlete_user_id: string;
  reason: string;
}

export async function submitErasureRequest(req: ErasureRequestIn): Promise<ErasureTicket> {
  return api<ErasureTicket>('/erasure', { method: 'POST', body: req });
}

export async function getErasureTicket(
  ticketId: string,
  opts: { signal?: AbortSignal } = {},
): Promise<ErasureTicket> {
  return api<ErasureTicket>(`/erasure/${ticketId}`, { signal: opts.signal });
}

export async function executeErasureTicket(ticketId: string): Promise<ErasureTicket> {
  return api<ErasureTicket>(`/erasure/${ticketId}/execute`, { method: 'POST' });
}
